test/source_test.go

```diff
@@ -1,20 +1,338 @@
 )
 
+var (
+	isRootLess     bool
+	isRootlessOnce sync.Once
+)
+
+func TestGomodGitAuthHTTPS(t *testing.T) {
+	t.Parallel()
+
+	// this is needed because go behaves differently when the host is
+	// recognized if it sees `host.docker.internal` it will first try to
+	// resolve the metadata by connecting to the resolved IP address on port
+	// 443, which will fail.
+	const host = "host.docker.internal"
+
+	ctx := startTestSpan(baseCtx, t)
+	sourceName := "gitauth"
+
+	// randomData := identity.NewID()
+
+	testEnv.RunTest(ctx, t, func(ctx context.Context, c gwclient.Client) {
+		const gomodFmt = `module %[1]s/user/public
+
+go 1.23.5
+
+require %[1]s/user/private.git v0.0.0
+`
+
+		gomodContents := fmt.Sprintf(gomodFmt, host)
+		port := getAvailablePort(t)
+
+		spec := &dalec.Spec{
+			Name: "gomod-git-auth",
+			Sources: map[string]dalec.Source{
+				sourceName: {
+					Inline: &dalec.SourceInline{
+						Dir: &dalec.SourceInlineDir{
+							Files: map[string]*dalec.SourceInlineFile{
+								"go.mod": {
+									Contents: gomodContents,
+								},
+							},
+						},
+					},
+					Generate: []*dalec.SourceGenerator{
+						{
+							Gomod: &dalec.GeneratorGomod{
+								Auth: map[string]dalec.GomodGitAuth{
+									fmt.Sprintf("%s:%d", host, port): {
+										Token: "super-secret",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		outsideAddr, insideAddr := getGitServerAddrs(ctx, t, c)
+		go runGitServer(outsideAddr, port)
+
+		sr := newSolveRequest(
+			withBuildTarget("debug/gomods"),
+			withSpec(ctx, t, spec),
+			withExtraHost(host, insideAddr),
+			withBuildContext(ctx, t, "gomod-worker", initGomodWorker(c, host, port)),
+		)
+		res := solveT(ctx, t, c, sr)
+
+		filename := filepath.Join(host, "user/private.git@v0.0.0/hello")
+		checkFile(ctx, t, filename, res, []byte("hello\n"))
+	}, testenv.WithSecrets(testenv.KeyVal{
+		K: "super-secret",
+		V: "value",
+	}))
+}
+
+func initGomodWorker(c gwclient.Client, host string, port int) llb.State {
+	worker := llb.Image("alpine:latest", llb.WithMetaResolver(c)).
+		Run(llb.Shlex("apk add --no-cache go git ca-certificates patch openssh")).Root()
+
+	// tell git to use the port along with the host
+	worker = worker.Run(
+		llb.Shlex(`sh -c 'git config --global "url.http://${HOST}:${PORT}/.insteadOf" "https://${HOST}"'`),
+		llb.AddEnv("HOST", host),
+		llb.AddEnv("PORT", fmt.Sprintf("%d", port)),
+	).Root()
+
+	// tell git to use the credential helper for the host:port combination
+	worker = worker.Run(
+		llb.Shlex(`sh -c 'git config --global credential."http://${HOST}:${PORT}.helper" "/usr/local/bin/frontend credential-helper --kind=token"'`),
+		llb.AddEnv("HOST", host),
+		llb.AddEnv("PORT", fmt.Sprintf("%d", port)),
+	).Root()
+	return worker
+}
+
+func getGitServerAddrs(ctx context.Context, t *testing.T, c gwclient.Client) (string, string) {
+	const (
+		rootlessOutsideAddr = "localhost"
+		rootlessInsideAddr  = "10.0.2.2" // as per the docs
```

Comment:
can you clarify which docs in your comment?

Reply:
Added to comment (done).